pkg/base1/test-import-json.ts

```diff
@@ -0,0 +1,264 @@
+
+let console_errors: string[] = [];
+const console_error = console.error;
+console.error = function (msg) {
+    console_errors.push(msg);
+    console_error(msg);
+};
+
+QUnit.hooks.beforeEach(() => {
+    console_errors = [];
+});
+
+QUnit.test("import_string", function(assert) {
+    assert.equal(import_string("foo"), "foo", "string");
+    assert.throws(() => import_string(12), "not a string");
+    assert.deepEqual(
+        console_errors,
+        [
+            'JSON validation error for : Not a string: 12'
+        ],
+        "console errors"
+    );
+});
+
+QUnit.test("import_number", function(assert) {
+    assert.equal(import_number(12), 12, "number");
+    assert.throws(() => import_number("foo"), "not a number");
+    assert.deepEqual(
+        console_errors,
+        [
+            'JSON validation error for : Not a number: "foo"'
+        ],
+        "console errors"
+    );
+});
+
+QUnit.test("import_boolean", function(assert) {
+    assert.equal(import_boolean(true), true, "boolean");
+    assert.throws(() => import_boolean("foo"), "not a boolean");
+    assert.deepEqual(
+        console_errors,
+        [
+            'JSON validation error for : Not a boolean: "foo"'
+        ],
+        "console errors"
+    );
+});
+
+QUnit.test("import_array", function(assert) {
+    assert.deepEqual(import_array(["a", "b", "c"], import_string), ["a", "b", "c"], "array of strings");
+    assert.deepEqual(import_array([1, 2, 3], import_number), [1, 2, 3], "array of numbers");
+    assert.deepEqual(import_array([1, 2, "c"], import_number), [1, 2], "array of numbers with a string");
+    assert.throws(() => import_array("foo", import_string), "not an array");
+    assert.deepEqual(
+        console_errors,
+        [
+            'JSON validation error for [2]: Not a number: "c"',
+            'JSON validation error for : Not an array: "foo"'
+        ],
+        "console errors"
+    );
+});
+
+QUnit.test("import_record", function(assert) {
+    assert.deepEqual(
+        import_record({ a: "a", b: "b", c: "c" }, import_string),
+        { a: "a", b: "b", c: "c" },
+        "record of strings");
+    assert.deepEqual(
+        import_record({ a: 1, b: 2, c: 3 }, import_number),
+        { a: 1, b: 2, c: 3 },
+        "record of numbers");
+    assert.deepEqual(
+        import_record({ a: 1, b: 2, c: "c" }, import_number),
+        { a: 1, b: 2 },
+        "record of numbers with a string");
+    assert.throws(() => import_record("foo", import_string), "not a record");
+    assert.deepEqual(
+        console_errors,
+        [
+            'JSON validation error for .c: Not a number: "c"',
+            'JSON validation error for : Not an object: "foo"'
+        ],
+        "console errors"
+    );
+});
+
+QUnit.test("validate", function(assert) {
+    assert.equal(validate("test input", "foo", import_string, ""), "foo", "string");
+    assert.equal(validate("test input", 12, import_string, ""), "", "not a string");
```

Comment:
Could you do "" → "default", to make it more obvious what that does?

Reply:
Yes.